src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -2405,16 +2487,43 @@ public void onChange(InetAddressAndPort endpoint, ApplicationState state, Versio
             {
                 switch (state)
                 {
+                    case RELEASE_VERSION:
+                        Nodes.nodes().update(endpoint, new CassandraVersion(value.value), NodeInfo::setReleaseVersion);
+                        break;
                     case DC:
+                        updateTopology(endpoint);
+                        Nodes.nodes().update(endpoint, value.value, NodeInfo::setDataCenter);
+                        break;
                     case RACK:
                         updateTopology(endpoint);
+                        Nodes.nodes().update(endpoint, value.value, NodeInfo::setRack);
+                        break;
+                    case RPC_ADDRESS:
+                    case NATIVE_ADDRESS_AND_PORT:
+                        try
+                        {
+                            Nodes.nodes().update(endpoint, InetAddressAndPort.getByName(value.value), NodeInfo::setNativeTransportAddressAndPort);
+                        }
+                        catch (UnknownHostException e)
+                        {
+                            throw new RuntimeException(e);
+                        }
+                        break;
+                    case SCHEMA:
+                        Nodes.nodes().update(endpoint, UUID.fromString(value.value), NodeInfo::setSchemaVersion);
+                        break;
+                    case HOST_ID:
+                        Nodes.nodes().update(endpoint, UUID.fromString(value.value), NodeInfo::setHostId);
                         break;
                     case RPC_READY:
                         notifyRpcChange(endpoint, epState.isRpcReady());
                         break;
                     case NET_VERSION:
                         updateNetVersion(endpoint, value);
                         break;
+                    case INDEX_STATUS:
+                        updateIndexStatus(endpoint, value);
+                        break;
```

Comment:
This is a correct place for this, however, note there is the same code repeated just above this switch in lines 2474-2477 (a result of STAR-540) and it makes no sense to keep both. I think you need to remove that snippet above and leave this one here in lines 2524-2526. Then it is also important that we add a case for INDEX_STATUS to `updatePeerInfo`, which is called from `handleStateNormal`. Otherwise the INDEX_STATUS information would be inaccurate shortly after joining the ring, because those statuses are ignored by `onChange` before the node becomes a member, and `handleStateNormal` is supposed to reapply them.